tests/lib.rs

```diff
@@ -661,3 +661,50 @@ fn arbitrary_impl_for_all_proto_variants() {
     let variants = core::mem::variant_count::<Protocol>() as u8;
     assert_eq!(variants, Proto::IMPL_VARIANT_COUNT);
 }
+
+#[test]
+fn multiaddr_with_p2p_and_from_str_and_peer() {
```

Comment:
Can you please split this into multiple tests:

- one for each happy path of each function
- one for each failure case

Reply:
It's done. Tell me if it answer what you wanted